# USA Gymnast with best chances to medal in Uneven Bars are:

- kaliya lincoln

In [1]:
import pandas as pd
import numpy as np
from scipy import stats
from mlxtend.preprocessing import minmax_scaling
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.colors import ListedColormap
from sklearn.model_selection import train_test_split


from sklearn.datasets import make_circles, make_classification, make_moons
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.ensemble import AdaBoostClassifier, RandomForestClassifier
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.inspection import DecisionBoundaryDisplay
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier


from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier


In [2]:
df = pd.read_csv('../Combine_Data/fx_encoded.csv')
olymp = pd.read_csv('../Data/cleandata22-23/encoded_w_olympics_fx.csv')
fxnames = pd.read_csv('../Data/cleandata22-23/encoded_w_olympics_fxnames.csv')
fxolymp = pd.read_csv('../Data/cleandata22-23/women22_23.csv')

In [3]:
df.columns

Index(['Rank', 'D Score', 'E Score', 'Pen.', 'Total', 'year', 'medal', 'Name',
       'nation', 'round_0', 'round_final', 'round_qual'],
      dtype='object')

In [4]:
# fx_encoded.csv
# fx_w.csv

In [5]:
df.head()

Rank  D Score  E Score  Pen.   Total  year  medal  Name  nation  round_0  \
0   1.0      6.2    8.266   0.0  14.466  2019      1   128      58    False   
1   2.0      5.6    7.833   0.0  13.433  2019      0   158      29    False   
2   3.0      5.2    8.100   0.0  13.300  2019      1   267      29    False   
3   4.0      5.4    7.933  -0.1  13.233  2019      0   192      42    False   
4   5.0      5.0    8.200   0.0  13.200  2019      0   191      19    False   

   round_final  round_qual  
0         True       False  
1         True       False  
2         True       False  
3         True       False  
4         True       False

In [6]:


clf = make_pipeline(StandardScaler(), SVC())

In [7]:
y = df['medal']
X = df.drop(columns=['medal'])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [8]:
clf = svm.SVC(decision_function_shape='ovo')
clf.fit(X_train, y_train)

SVC(decision_function_shape='ovo')

In [9]:
dec = clf.decision_function(X_test)

In [10]:
dec.shape

(88,)

In [11]:
clf.decision_function_shape = "ovr"

In [12]:

# Initialize the SVM classifier
clf = SVC(decision_function_shape='ovo')

# Fit the model to the training data
clf.fit(X_train, y_train)

# Make predictions on the test data
y_pred = clf.predict(X_test)

# Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy)
print("Classification Report:\n", report)
print("Confusion Matrix:\n", conf_matrix)


Accuracy: 0.9545454545454546
Classification Report:
               precision    recall  f1-score   support

           0       0.95      1.00      0.98        84
           1       0.00      0.00      0.00         4

    accuracy                           0.95        88
   macro avg       0.48      0.50      0.49        88
weighted avg       0.91      0.95      0.93        88

Confusion Matrix:
 [[84  0]
 [ 4  0]]


/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, m

In [13]:

# Assuming you already have your DataFrame 'df' and the target variable is 'medal'
y = df['medal']
X = df.drop(columns=['medal'])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize classifiers
classifiers = {
    "Random Forest": RandomForestClassifier(),
    "AdaBoost": AdaBoostClassifier(),
    "SVM": SVC(),
    "K-Nearest Neighbors": KNeighborsClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "Naive Bayes": GaussianNB(),
    "Neural Network": MLPClassifier(),
}

results = {}

# Iterate through each classifier
for name, clf in classifiers.items():
    # Create a pipeline with StandardScaler for classifiers that require it
    if name in ["SVM", "K-Nearest Neighbors", "Neural Network"]:
        clf = make_pipeline(StandardScaler(), clf)

    # Fit the model to the training data
    clf.fit(X_train, y_train)

    # Make predictions on the test data
    y_pred = clf.predict(X_test)

    # Evaluate the model's performance
    accuracy = accuracy_score(y_test, y_pred)
    report = classification_report(y_test, y_pred)
    conf_matrix = confusion_matrix(y_test, y_pred)

    results[name] = {
        "accuracy": accuracy,
        "classification_report": report,
        "confusion_matrix": conf_matrix,
    }

# Print the results
for name, result in results.items():
    print(f"Classifier: {name}")
    print(f"Accuracy: {result['accuracy']:.4f}")
    print(f"Classification Report:\n{result['classification_report']}")
    print(f"Confusion Matrix:\n{result['confusion_matrix']}\n")


/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, m

Classifier: Random Forest
Accuracy: 0.9659
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98        84
           1       1.00      0.25      0.40         4

    accuracy                           0.97        88
   macro avg       0.98      0.62      0.69        88
weighted avg       0.97      0.97      0.96        88

Confusion Matrix:
[[84  0]
 [ 3  1]]

Classifier: AdaBoost
Accuracy: 0.9659
Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98        84
           1       1.00      0.25      0.40         4

    accuracy                           0.97        88
   macro avg       0.98      0.62      0.69        88
weighted avg       0.97      0.97      0.96        88

Confusion Matrix:
[[84  0]
 [ 3  1]]

Classifier: SVM
Accuracy: 0.9545
Classification Report:
              precision    recall  f1-score   support

           0       0.95      1.0

/Users/ryantalbot/opt/anaconda3/envs/tf2/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Classifier: Random Forest
Accuracy: 0.9659

Classifier: SVM
Accuracy: 0.9545

Classifier: K-Nearest Neighbors
Accuracy: 0.9659

Classifier: Neural Network
Accuracy: 0.9659

Classifier: AdaBoost
Accuracy: 0.9659

In [14]:

# Assuming you already have your DataFrame 'df' and the target variable is 'medal'
y = df['medal']
X = df.drop(columns=['medal'])

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the AdaBoost classifier
clf = AdaBoostClassifier(n_estimators=100,  random_state=42)

# Fit the model to the training data
clf.fit(X_train, y_train)

# Make predictions on the test data
y_pred = clf.predict(X_test)

# Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)

# Print the results
print("AdaBoost Classifier")
print(f"Accuracy: {accuracy:.4f}")
print(f"Classification Report:\n{report}")
print(f"Confusion Matrix:\n{conf_matrix}")


AdaBoost Classifier
Accuracy: 0.9886
Classification Report:
              precision    recall  f1-score   support

           0       0.99      1.00      0.99        84
           1       1.00      0.75      0.86         4

    accuracy                           0.99        88
   macro avg       0.99      0.88      0.93        88
weighted avg       0.99      0.99      0.99        88

Confusion Matrix:
[[84  0]
 [ 1  3]]


In [15]:
X_train.head()

Rank  D Score  E Score  Pen.   Total  year  Name  nation  round_0  \
265   7.0      6.1    7.633   0.0  13.733  2013   236      47    False   
406  48.0      4.6    7.466   0.0  12.066  2015   101      24    False   
31   24.0      4.6    6.900   0.0  11.500  2019    71      56    False   
84    5.0      5.2    8.450   0.0  13.650  2017    48       0    False   
378  20.0      5.7    7.533   0.0  13.233  2015   193      29    False   

     round_final  round_qual  
265         True       False  
406        False        True  
31         False        True  
84          True       False  
378        False        True

In [16]:
olymp.head()


Rank  D Score  E Score  Pen.   Total  year  Name  nation  round_aafinal  \
0  109.0      4.4    7.466   0.1  11.766  2022   227      22          False   
1   22.0      4.5    7.366   0.0  11.866  2023   227      22          False   
2   39.0      2.9    4.450   0.0   7.350  2022   286      67          False   
3   74.0      4.4    7.866   0.0  12.266  2022   228      22          False   
4   16.0      5.3    8.033   0.0  13.333  2022   414      28          False   

   round_final  round_qual  round_teamfinal  
0        False        True            False  
1        False        True            False  
2        False        True            False  
3        False        True            False  
4        False        True            False

In [17]:
olymp=olymp.drop(columns=['round_aafinal', 'round_teamfinal'])

In [18]:
olymp['round_0']=False

In [19]:
olymp.columns

Index(['Rank', 'D Score', 'E Score', 'Pen.', 'Total', 'year', 'Name', 'nation',
       'round_final', 'round_qual', 'round_0'],
      dtype='object')

In [20]:
# Define the desired column order
desired_order = ['Rank', 'D Score', 'E Score', 'Pen.', 'Total', 'year', 'Name', 'nation',
       'round_0', 'round_final', 'round_qual']

# Reorder the columns based on the desired order
olymp = olymp[desired_order]

In [21]:
y_pred = clf.predict(olymp)


In [22]:
y_pred

array([0, 0, 0, ..., 0, 0, 0])

In [23]:
ypreds  = pd.Series(y_pred)

In [24]:
olymp['ypred']=ypreds

In [25]:
olymp['ypred'].value_counts()

ypred
0    1214
1      10
Name: count, dtype: int64

In [26]:
olymp

Rank  D Score  E Score  Pen.   Total  year  Name  nation  round_0  \
0     109.0      4.4    7.466   0.1  11.766  2022   227      22    False   
1      22.0      4.5    7.366   0.0  11.866  2023   227      22    False   
2      39.0      2.9    4.450   0.0   7.350  2022   286      67    False   
3      74.0      4.4    7.866   0.0  12.266  2022   228      22    False   
4      16.0      5.3    8.033   0.0  13.333  2022   414      28    False   
...     ...      ...      ...   ...     ...   ...   ...     ...      ...   
1219    6.0      5.2    7.950   0.0  13.150  2023   305      76    False   
1220    0.0      4.9    7.800   0.0  12.700  2023    22      76    False   
1221    0.0      4.9    8.000   0.0  12.900  2023    22      76    False   
1222   13.0      4.9    7.650   0.0  12.550  2023    22      76    False   
1223   20.0      4.3    7.033   0.0  11.333  2023   285      81    False   

      round_final  round_qual  ypred  
0           False        True      0  
1           False        True      0  
2           False        True      0  
3           False        True      0  
4           False        True      0  
...           ...         ...    ...  
1219        False       False      0  
1220        False        True      0  
1221         True       False      0  
1222        False       False      0  
1223        False        True      0  

[1224 rows x 12 columns]

In [27]:
filtered_df = olymp[olymp['ypred'] == 1]

In [28]:
filtered_df

Rank  D Score  E Score  Pen.  Total  year  Name  nation  round_0  \
327   1.0      6.0     8.40   0.0  14.40  2022   241      28    False   
328   1.0      6.0     8.20   0.0  14.20  2022   241      28    False   
330   1.0      5.8     8.80   0.0  14.60  2022   241      28    False   
332   1.0      5.7     8.30   0.0  14.00  2022   241      28    False   
336   1.0      5.8     8.20   0.0  14.00  2023   241      28    False   
337   1.0      5.8     8.10   0.0  13.90  2023   241      28    False   
603   0.0      6.1     8.25   0.0  14.35  2023   258      76    False   
711   1.0      5.3     8.25   0.0  13.55  2023    15      50    False   
712   1.0      5.3     8.05   0.0  13.35  2023    15      50    False   
950   1.0      5.8     8.40   0.0  14.20  2022   185       8    False   

     round_final  round_qual  ypred  
327        False       False      1  
328         True       False      1  
330         True       False      1  
332         True       False      1  
336         True       False      1  
337        False        True      1  
603        False        True      1  
711         True       False      1  
712        False        True      1  
950        False        True      1

In [29]:
filtered_df['Name'].unique()

array([241, 258,  15, 185])

In [30]:
fxnames[fxnames['Name_encoded']==241]

Name  Name_encoded
327  jessica gadirova           241
328  jessica gadirova           241
329  jessica gadirova           241
330  jessica gadirova           241
331  jessica gadirova           241
332  jessica gadirova           241
333  jessica gadirova           241
334  jessica gadirova           241
335  jessica gadirova           241
336  jessica gadirova           241
337  jessica gadirova           241

In [31]:
fxnames[fxnames['Name_encoded']==258]

Name  Name_encoded
603  kaliya lincoln           258
604  kaliya lincoln           258
605  kaliya lincoln           258

In [32]:
fxnames[fxnames['Name_encoded']==15]

Name  Name_encoded
710  alexa moreno            15
711  alexa moreno            15
712  alexa moreno            15

In [33]:
fxnames[fxnames['Name_encoded']==185]

Name  Name_encoded
950  flavia saraiva           185

# Medalists for womens floor event (in no particular order)
- jessica gadirova
- kaliya lincoln
- alexa moreno
- flavia saraiva

In [36]:
# List of Name_encoded values you want to filter by
name_encoded_values = [241, 258,  15, 185]

# Create an empty set to store unique names
unique_names = set()

# Iterate through the DataFrame and add unique names to the set
for index, row in fxnames.iterrows():
    if row['Name_encoded'] in name_encoded_values:
        unique_names.add(row['Name'])

# Convert the set of unique names back to a list
unique_names_list = list(unique_names)


In [37]:
unique_names_list

['jessica gadirova', 'flavia saraiva', 'kaliya lincoln', 'alexa moreno']

In [38]:
fxolymp['Name'] = fxolymp['Name'].str.lower()

In [39]:
names_to_find = ['jessica gadirova', 'flavia saraiva', 'kaliya lincoln', 'alexa moreno']

# Filter the DataFrame to get the desired names and their corresponding countries
result_df = fxolymp[fxolymp['Name'].isin(names_to_find)]


In [40]:
result_df=result_df[result_df['Apparatus']== 'FX']

In [41]:
usa = result_df[result_df['Country']=='USA']

In [43]:
usa['Name'].unique()

array(['kaliya lincoln'], dtype=object)